classes/submission/SubmissionDAO.inc.php

```diff
@@ -94,30 +95,110 @@ function newDataObject() {
 	/**
 	 * Internal function to return a Submission object from a row.
 	 * @param $row array
+	 * @param $submissionRevision int optional
 	 * @return Submission
 	 */
-	function _fromRow($row) {
+	function _fromRow($row, $submissionRevision = 1) {
 		$submission = $this->newDataObject();
 
 		$submission->setId($row['submission_id']);
+		$submission->setSubmissionRevision($submissionRevision);
 		$submission->setContextId($row['context_id']);
 		$submission->setLocale($row['locale']);
 		$submission->setStageId($row['stage_id']);
 		$submission->setStatus($row['status']);
 		$submission->setSubmissionProgress($row['submission_progress']);
 		$submission->setDateSubmitted($this->datetimeFromDB($row['date_submitted']));
 		$submission->setDateStatusModified($this->datetimeFromDB($row['date_status_modified']));
-		$submission->setDatePublished($this->datetimeFromDB($row['date_published']));
 		$submission->setLastModified($this->datetimeFromDB($row['last_modified']));
 		$submission->setLanguage($row['language']);
 		$submission->setCitations($row['citations']);
 
-		$this->getDataObjectSettings('submission_settings', 'submission_id', $submission->getId(), $submission);
+		$this->getDataObjectSettings('submission_settings', 'submission_id', $submission->getId(), $submission, $submissionRevision, 'submission_revision');
 
 		return $submission;
 	}
 
 	/**
+	 * Get all revisions for a submission
+	 * @param $submissionId int
+	 * @param $contextId int optional
+	 * $param $order string optional default: ASC
+	 * @return array
+	 */
+	function getSubmissionRevisionIds($submissionId, $contextId = null, $order = SORT_DIRECTION_ASC) {
+
+		if (!ctype_digit("$submissionId")) {
+			$publishedArticleDao = DAORegistry::getDAO('PublishedArticleDAO');
+			$publishedArticle = $publishedArticleDao->getPublishedArticleByPubId('publisher-id', $submissionId, $contextId);
+			$submissionId = $publishedArticle->getId();
+		}
+
+		$params = array((int) $submissionId);
+		if ($contextId) {
+			$params[] = (int) $contextId;
+		}
+
+		$sql = 'SELECT DISTINCT submission_revision FROM submission_settings ' .
+			   'INNER JOIN submissions ON submission_settings.submission_id = submissions.submission_id' .
+			   ' WHERE submission_settings.submission_id = ?'  .
+			   ($contextId ? ' AND submissions.context_id = ?' : '') .
+			   ' ORDER BY submission_revision ' . $this->getDirectionMapping($order);
+
+		$result = $this->retrieve($sql, $params);
+		$submissionRevisions = array();
+
+		foreach($result->getArray() as $submission) {
+			$submissionRevisions[] = $submission['submission_revision'];
+		}
+
+		return $submissionRevisions;
+	}
+
+	/**
+	 * Get all published revisions for a submission
+	 * @param $submissionId int
+	 * @return array
+	*/
+	function getPublishedSubmissionRevisions($submissionId, $contextId = null, $order = SORT_DIRECTION_DESC){
```

Comment:
because this considers only published submissions, it should probably be places in the PublishedArticleDAO?